File: shared/athena-lms-common/src/main/java/com/athena/lms/common/event/DomainEvent.java
package com.athena.lms.common.event;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.Instant;
import java.util.UUID;

/**
 * Generic domain event envelope for all LMS services.
 * Published to athena.lms.exchange with routing key = event type (e.g. "loan.disbursed").
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class DomainEvent<T> {

    /** Unique event ID (UUID). */
    @Builder.Default
    private String id = UUID.randomUUID().toString();

    /** Routing key / event type (e.g. "loan.disbursed", "payment.completed"). */
    private String type;

    /** Schema version for forward compatibility. */
    @Builder.Default
    private int version = 1;

    /** Originating service name. */
    private String source;

    /** Tenant identifier for multi-tenant routing. */
    private String tenantId;

    /** Correlation ID for distributed tracing. */
    private String correlationId;

    @Builder.Default
    @JsonFormat(shape = JsonFormat.Shape.STRING)
    private Instant timestamp = Instant.now();

    /** Domain-specific payload. */
    private T payload;

    // ─── Convenience factory ──────────────────────────────────────────────────

    public static <T> DomainEvent<T> of(String type, String source, String tenantId, T payload) {
        return DomainEvent.<T>builder()
                .type(type)
                .source(source)
                .tenantId(tenantId)
                .payload(payload)
                .build();
    }
}

File: go-services/internal/common/event/domain_event.go
package event

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// DomainEvent is the generic event envelope for all LMS services.
// Published to athena.lms.exchange with routing key = event type.
// JSON shape must be byte-compatible with the Java DomainEvent<T>.
type DomainEvent struct {
	// Unique event ID (UUID).
	ID string `json:"id"`

	// Routing key / event type (e.g. "loan.disbursed").
	Type string `json:"type"`

	// Schema version for forward compatibility.
	Version int `json:"version"`

	// Originating service name.
	Source string `json:"source"`

	// Tenant identifier for multi-tenant routing.
	TenantID string `json:"tenantId"`

	// Correlation ID for distributed tracing.
	CorrelationID string `json:"correlationId,omitempty"`

	// ISO-8601 timestamp — matches Java Instant serialization.
	Timestamp string `json:"timestamp"`

	// Domain-specific payload (raw JSON to preserve type flexibility).
	Payload json.RawMessage `json:"payload"`
}

// NewDomainEvent creates a new DomainEvent with auto-populated ID and timestamp.
func NewDomainEvent(eventType, source, tenantID, correlationID string, payload any) (*DomainEvent, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, err
	}

	return &DomainEvent{
		ID:            uuid.New().String(),
		Type:          eventType,
		Version:       1,
		Source:        source,
		TenantID:      tenantID,
		CorrelationID: correlationID,
		Timestamp:     time.Now().UTC().Format(time.RFC3339Nano),
		Payload:       payloadBytes,
	}, nil
}

// UnmarshalPayload decodes the Payload into the given target.
func (e *DomainEvent) UnmarshalPayload(target any) error {
	return json.Unmarshal(e.Payload, target)
}
